File: easy-projects/easy-common/src/main/java/xyz/srclab/common/lang/computed/Computed.java
package xyz.srclab.common.lang.computed;

import xyz.srclab.annotation.Immutable;

import java.time.Duration;
import java.util.function.Supplier;

/**
 * Caches a computed value, particularly in cases the value is optional and has a heavy process to get.
 *
 * @param <T> computed value
 */
@Immutable
public interface Computed<T> {

    static <T> Computed<T> onceOf(Supplier<T> computation) {
        return ComputedSupport.newOnceComputed(computation);
    }

    static <T> Computed<T> multiOf(Supplier<Computed.Result<T>> computation) {
        return ComputedSupport.newMultiComputed(computation);
    }

    static <T> Computed<T> refreshableOf(long timeoutNanos, Supplier<T> computation) {
        return ComputedSupport.newRefreshableComputed(timeoutNanos, computation);
    }

    static <T> Computed<T> refreshableOf(Duration duration, Supplier<T> computation) {
        return ComputedSupport.newRefreshableComputed(duration, computation);
    }

    T get();

    T refreshAndGet();

    interface Result<T> {

        T value();

        Duration expiry();
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/base/Computed.kt
package xyz.srclab.common.base

import java.time.Duration
import java.util.function.Supplier

/**
 * Caches a computed value, particularly in cases the value is optional and has a heavy process to get.
 *
 * @param <T> computed value
 */
interface Computed<T> {

    fun get(): T

    fun refresh()

    fun refreshAndGet(): T

    interface Result<T> {
        fun value(): T
        fun expiry(): Duration?
    }

    companion object {
        fun <T> onceOf(computation: Supplier<T>?): Computed<T>? {
            return ComputedSupport.newOnceComputed(computation)
        }

        fun <T> multiOf(computation: Supplier<Result<T>?>?): Computed<T>? {
            return ComputedSupport.newMultiComputed(computation)
        }

        fun <T> refreshableOf(timeoutNanos: Long, computation: Supplier<T>?): Computed<T>? {
            return ComputedSupport.newRefreshableComputed(timeoutNanos, computation)
        }

        fun <T> refreshableOf(duration: Duration?, computation: Supplier<T>?): Computed<T>? {
            return ComputedSupport.newRefreshableComputed(duration, computation)
        }
    }
}